pwndbg/heap/ptmalloc.py

```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+                            self._main_arena_addr = base_offset + instr.disp # eax + disp
+                            break
+                elif pwndbg.arch.current == "aarch64":
+                    base_offset = pwndbg.vmmap.find(pwndbg.symbol.address('_IO_list_all')).start
+                    reg = None
+                    for instr in malloc_trim_instructions[5:]:
+                        # Try to find `add reg2, reg1, #offset` after `adrp reg1, #base_offset`
+                        if instr.mnemonic == 'add' and instr.operands[1].str == reg:
+                            self._main_arena_addr = base_offset + instr.operands[2].int
+                            break
+                        if instr.mnemonic == 'adrp' and instr.operands[1].int == base_offset:
+                            reg = instr.operands[0].str
+                elif pwndbg.arch.current == "arm":
+                    ldrw_instr = None
+                    for instr in malloc_trim_instructions:
+                        # Try to find `ldr.w reg, [pc, #offset]`, then `add reg, pc`
+                        if not ldrw_instr:
+                            if instr.mnemonic == 'ldr.w':
+                                ldrw_instr = instr
+                        else:
+                            reg = ldrw_instr.operands[0].str
+                            if instr.mnemonic == 'add' and instr.op_str == reg + ", pc":
+                                # ldr.w reg, [pc, #offset]
+                                offset = ldrw_instr.operands[1].mem.disp
+                                offset = pwndbg.memory.s32((ldrw_instr.address + 4 & -4) + offset)
+                                # add reg, pc
+                                self._main_arena_addr = offset + instr.address + 4
+                else:
+                    inform_report_issue("main_arena")
+                    raise OSError("Cannot find the symbol via heuristics")
+            # try to search main_arena in .data of libc if we can't find it via above trick
+            if not self._main_arena_addr:
+                _IO_2_1_stdin_addr = pwndbg.symbol.address('_IO_2_1_stdin_')
+                _IO_list_all_addr = pwndbg.symbol.address('_IO_list_all')
+                # main_arena is between _IO_2_1_stdin and _IO_list_all
+                for addr in range(_IO_2_1_stdin_addr, _IO_list_all_addr, pwndbg.arch.ptrsize):
+                    tmp_arena = self.malloc_state(addr)
+                    if tmp_arena["next"] == addr:
+                        self._main_arena_addr = addr
+                        break
+                if not self._main_arena_addr:
+                    # there are more than one arena, try to find by main_arena.top and main_arena.max_system_mem
+                    heap_page = next(x for x in pwndbg.vmmap.get() if "heap]" in x.objfile)
+                    for addr in range(_IO_2_1_stdin_addr, _IO_list_all_addr, pwndbg.arch.ptrsize):
```

Comment:
How does `main_arena->next` work? I mean, I can see that when there is only one arena, the `main_arena->next == main_arena`. If we have two arenas, is it cyclic like `main_arena->next->next == main_arena`? If so, why don't we try to resolve it like this?

Reply:
Good idea!
I didn't think about this before, I'll try to support what you said.